src/sampling.jl

```diff
@@ -0,0 +1,177 @@
+using StatsBase
+"""
+	Contains all function for sampling expressions and from expressions
+"""
+
+"""
+    rand(::Type{RuleNode}, grammar::Grammar, typ::Symbol, max_depth::Int=10)
+
+Generates a random RuleNode of return type typ and maximum depth max_depth.
+"""
+function Base.rand(::Type{RuleNode}, grammar::Grammar, typ::Symbol, max_depth::Int=10, 
+    bin::Union{NodeRecycler,Nothing}=nothing)
+    rules = grammar[typ]
```

Comment:
add constraint propagation as in here:     https://github.com/Herb-AI/HerbSearch.jl/blob/debbc00ebc42782595539d9a105f74ccea25a418/src/csg_priority_enumerator.jl#L27

Reply:
Maybe it is also a good idea to move the sampling functions to the search.
If we leave it in the HerbGrammar, we get a cyclic dependency between HerbGrammar and HerbConstraints, since you would have to import HerbConstraints to get propagation to work.

In `csg_priority_enumerator.jl` in HerbSearch, there is a `propagate_constraints` function that you can use to filter a list of possible rules.

When constraints are being propagated, we need a `GrammarContext`, which contains:
 - A list of `LocalConstraint`s. You have to keep track of all `LocalConstraint`s for a tree that is being sampled (one list for the entire tree). The `propagate` function takes the local constraints and returns an updated list of local constraints. Note that it doesn't update the same list, but returns a new list.
 - The actual tree that is being expanded. I think this can be achieved by just passing it in the recursive calls as a reference.
 - A path from the root to the hole that is currently being expanded (i.e. the node that you are sampling). A path is given by a sequence of child indices, e.g. [1, 2] is the second child of the first child of the root. 

So, you do still need to modify this function a bit to make it work with constraint propagation.
Let me know if you get stuck or need help!